extensions/standard-processors/processors/AttributeRollingWindow.cpp

```diff
@@ -0,0 +1,121 @@
+  if (!current_value_opt) {
+    logger_->log_warn("Missing value to track, flow file uuid: {}", flow_file->getUUIDStr());
+    session->transfer(flow_file, Failure);
+    return;
+  }
+  const auto current_value = [&current_value_opt] {
+    try {
+      return std::stod(*current_value_opt);
+    } catch (const std::exception& ex) {
+      throw minifi::Exception{ExceptionType::PROCESSOR_EXCEPTION,
+          fmt::format("Failed to convert 'Value to track' of '{}' to double", *current_value_opt)};
+    }
+  }();
+  // copy: so we can release the lock sooner
+  const auto state_copy = [&, now = std::chrono::system_clock::now()] {
+    const std::lock_guard lg{state_mutex_};
+    state_.add(now, current_value);
+    if (window_length_) {
+      state_.shrinkToSize(*window_length_);
+    } else {
+      gsl_Assert(time_window_);
+      state_.removeOlderThan(now - *time_window_);
+    }
+    return state_.getEntries();
+  }();
+  const auto sorted_values = [&state_copy] {
+    auto values = state_copy | ranges::views::transform(&decltype(state_)::Entry::value) | ranges::to<std::vector>;
+    std::sort(std::begin(values), std::end(values));
+    return values;
+  }();
+  calculateAndSetAttributes(*flow_file, sorted_values);
+  session->transfer(flow_file, Success);
+}
+
+/**
+ * Calculate statistical properties of the values in the rolling window and set them as attributes on the flow file.
+ * Properties: count, value (sum), mean (average), median, variance, stddev
+ */
+void AttributeRollingWindow::calculateAndSetAttributes(core::FlowFile &flow_file,
+    std::span<const double> sorted_values) const {
+  const auto attribute_name = [this](std::string_view suffix) {
+    return utils::string::join_pack(attribute_name_prefix_, suffix);
+  };
+  const auto set_aggregate = [&flow_file, attribute_name](std::string_view name, double value) {
+    flow_file.setAttribute(attribute_name(name), std::to_string(value));
+  };
+  set_aggregate("count", sorted_values.size());
+  const auto sum = std::accumulate(std::begin(sorted_values), std::end(sorted_values), 0.0);
+  set_aggregate("value", sum);
+  const auto mean = sum / gsl::narrow_cast<double>(sorted_values.size());
+  set_aggregate("mean", mean);
+  set_aggregate("median", [&] {
+    const auto mid = sorted_values.size() / 2;
+    return sorted_values.size() % 2 == 0
+        ? std::midpoint(sorted_values[mid], sorted_values[mid - 1])  // even number of values: average the two middle values
+        : sorted_values[mid];  // odd number of values: take the middle value
+  }());
+  // https://math.stackexchange.com/questions/1720876/sums-of-squares-minus-square-of-sums
+  const auto avg_of_squares = std::accumulate(std::begin(sorted_values), std::end(sorted_values), 0.0, [&](double acc, double value) {
+    return acc + std::pow(value, 2) / gsl::narrow_cast<double>(sorted_values.size());
+  });
+  const auto variance = avg_of_squares - std::pow(mean, 2);
```

Comment:
It may be faster, but it's more prone to overflow when calculating the average of large numbers. I'm not sure how much of a problem possible overflow is with floating point numbers, but this stackoverflow answer also recommends dividing in each step: https://stackoverflow.com/a/1934266/3997716